### PR TITLE
v 1.0

### PR DESCRIPTION
**Pull request by Anton-Zhytonbaiev**
[DEMO LINK](https://Anton-Zhytonbaiev.github.io/react_todo-app-with-api/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,245 @@
...
+          setTodoTitleWrapper={setTodoTitleWrapper}
+          setErrorWrapper={setErrorWrapper}
+          addNewTodo={addNewTodo}
+          isLoading={isAddWaiting}
+          isAllTodosCompleted={isAllTodosCompleted}
+          haveTodos={haveTodos}
+          toggleAllTodosStatus={toggleAllTodosStatus}
+        />
+
+        {(todos.length > 0 || tempTodo !== null) && (
```
Comment (vanst1) [inline src/App.tsx:208]: let's create variable for this

File: src/components/Errors/Errors.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        'Unable to add a todo'
+      )}
+
+      {error === Error.ONDELETE && (
+        'Unable to delete a todo'
+      )}
+
+      {error === Error.ONUPDATE && (
+        'Unable to update a todo'
+      )}
```
Comment (vanst1) [inline src/components/Errors/Errors.tsx:61]: create some function for the generation error messages

**Review by vanst1 — CHANGES_REQUESTED** [review]
Good job) Some fix needed:)

Create `index.tsx` for each component)

File: src/App.tsx [hunk 3]
```diff
@@ -1,24 +1,248 @@
...
+
+  const changeRemoveTodoIds = useCallback((value: number[]) => {
+    setOnRemoveTodoIds(value);
+  }, []);
+
+  const removeError = () => {
+    window.setTimeout(() => setError(Error.NONE), 3000);
+  };
+
+  const loadTodosFromServer = async () => {
```
Comment (ArturErmolenko) [inline src/App.tsx:42]: useCallback

File: src/App.tsx [hunk 4]
```diff
@@ -1,24 +1,248 @@
...
+
+  const setTodoTitleWrapper = (value: string) => {
+    setTodoTitle(value);
+  };
+
+  const setErrorWrapper = (currentError: Error) => {
+    setError(currentError);
+  };
+
+  const completedTodoIds = completedTodos.map(todo => todo.id);
```
Comment (ArturErmolenko) [inline src/App.tsx:154]: useMemo

File: src/components/Errors/Errors.tsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import classNames from 'classnames';
+import { Error } from '../../types/Error';
+
+type Props = {
+  error: Error,
+  setErrorWrapper: (value: Error) => void;
+  removeError: () => void;
+};
+
+export const Errors: React.FC<Props> = ({
```
Comment (ArturErmolenko) [inline src/components/Errors/Errors.tsx:11]: memo

File: src/components/Errors/Errors.tsx (hunk 2 above)
Comment (ArturErmolenko) [reply src/components/Errors/Errors.tsx:61]: I strongly agree

File: src/components/Header/Header.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  setTodoTitleWrapper: (value: string) => void;
+  setErrorWrapper: (error: Error) => void;
+  addNewTodo: () => void;
+  isLoading: boolean;
+  isAllTodosCompleted: boolean;
+  toggleAllTodosStatus: (status: boolean) => void;
+  haveTodos: boolean;
+};
+
+export const Header: React.FC<Props> = ({
```
Comment (ArturErmolenko) [inline src/components/Header/Header.tsx:16]: memo

File: src/components/TodoItem/TodoItem.tsx [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+  todosIdsToRemove: number[];
+  changeCompletedStatus: (todoId: number, status: boolean) => void;
+  isUpdateWaiting: boolean;
+  todosIdsToUpdate: number[];
+  changeTodosIdsToUpdate: (value: number) => void;
+  removeUpdatedId: (value: number) => void;
+  changeTodoTitle: (todoId: number, newTitle: string) => void;
+};
+
+export const TodoItem: React.FC<Props> = ({
```
Comment (ArturErmolenko) [inline src/components/TodoItem/TodoItem.tsx:24]: memo

File: src/components/TodoList/TodoList.tsx [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  removeDeleteId: (id: number) => void;
+  changeCompletedStatus: (todoId: number, status: boolean) => void;
+  isUpdateWaiting: boolean;
+  todosIdsToUpdate: number[];
+  changeTodosIdsToUpdate: (value: number) => void;
+  removeUpdatedId: (value: number) => void;
+  changeTodoTitle: (todoId: number, newTitle: string) => void;
+};
+
+export const TodoList: React.FC<Props> = ({
```
Comment (ArturErmolenko) [inline src/components/TodoList/TodoList.tsx:21]: memo

File: src/components/TodoList/TodoList.tsx [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  changeRemoveTodoIds,
+  removeDeleteId,
+  changeCompletedStatus,
+  isUpdateWaiting,
+  todosIdsToUpdate,
+  changeTodosIdsToUpdate,
+  removeUpdatedId,
+  changeTodoTitle,
+}) => {
+  return (
```
Comment (ArturErmolenko) [inline src/components/TodoList/TodoList.tsx:36]: redundant return

**Review by ArturErmolenko — CHANGES_REQUESTED** [review]

**Review by ArturErmolenko — APPROVED** [review]
Well done 🔥

**Pull request closed without merging**